From f3f592d14b0362933aeeb5d7d891193529d8ce0a Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Sat, 14 Dec 2013 15:50:37 +0000
Subject: [PATCH] builder: Add a test of the planner.

This tries out many combinations of size/format and ensures
that the planner can deal with them.

(cherry picked from commit a27963b67ba1140c29c145a1ce14dbfc05f8d7da)
---
 .gitignore                           |  1 +
 builder/Makefile.am                  | 24 ++++++++++++++---
 builder/test-index.in                | 26 ++++++++++++++++++
 builder/test-virt-builder-planner.sh | 51 ++++++++++++++++++++++++++++++++++++
 4 files changed, 98 insertions(+), 4 deletions(-)
 create mode 100755 builder/test-virt-builder-planner.sh

diff --git a/.gitignore b/.gitignore
index 43f9f40..b967dbd 100644
--- a/.gitignore
+++ b/.gitignore
@@ -61,6 +61,7 @@ Makefile.in
 /builder/index-parse.c
 /builder/index-parse.h
 /builder/index-scan.c
+/builder/*.qcow2
 /builder/stamp-virt-builder.pod
 /builder/stamp-virt-index-validate.pod
 /builder/test-index
diff --git a/builder/Makefile.am b/builder/Makefile.am
index ca4a254..11967e9 100644
--- a/builder/Makefile.am
+++ b/builder/Makefile.am
@@ -143,15 +143,28 @@ CLEANFILES += stamp-virt-builder.pod
 
 TESTS_ENVIRONMENT = $(top_builddir)/run --test
 
-all_disk_images := debian.xz fedora.xz ubuntu.xz windows.xz
-disk_images := $(shell for f in $(all_disk_images); do b=`basename $$f .xz`; if [ -f "../tests/guests/$$b.img" ]; then echo $$f; fi; done)
+disk_images := \
+	$(shell for f in debian fedora ubuntu windows; do if [ -s "../tests/guests/$$f.img" ]; then echo $$f.xz; fi; done) \
+	$(shell if [ -s "../tests/guests/fedora.img" ]; then echo fedora.qcow2 fedora.qcow2.xz; fi)
 
-CLEANFILES += $(all_disk_images)
+CLEANFILES += *.qcow2 *.xz
 
 check_DATA = $(disk_images)
 
+fedora.qcow2: ../tests/guests/fedora.img
+	rm -f $@ $@-t
+	qemu-img convert -f raw -O qcow2 $< $@-t
+	mv $@-t $@
+
+fedora.qcow2.xz: fedora.qcow2
+	rm -f $@ $@-t
+	xz --best --block-size=16777216 -c $< > $@-t
+	mv $@-t $@
+
 %.xz: ../tests/guests/%.img
-	xz --best --block-size=16777216 -c $< > $@
+	rm -f $@ $@-t
+	xz --best --block-size=16777216 -c $< > $@-t
+	mv $@-t $@
 
 TESTS = test-virt-builder-list.sh
 
@@ -162,6 +175,9 @@ endif ENABLE_APPLIANCE
 check-valgrind:
 	$(MAKE) VG="$(top_builddir)/run @VG@" check
 
+check-slow:
+	$(MAKE) TESTS="test-virt-builder-planner.sh" check
+
 # Dependencies.
 depend: .depend
 
diff --git a/builder/test-index.in b/builder/test-index.in
index 21e2ca3..1bca6b8 100644
--- a/builder/test-index.in
+++ b/builder/test-index.in
@@ -16,6 +16,32 @@ expand=/dev/sda2
 lvexpand=/dev/VG/Root
 notes=Phony Fedora look-alike used for testing.
 
+[phony-fedora-qcow2]
+name=Phony Fedora qcow2
+file=fedora.qcow2.xz
+format=qcow2
+size=1073741824
+expand=/dev/sda2
+lvexpand=/dev/VG/Root
+notes=Phony Fedora look-alike used for testing.
+
+[phony-fedora-qcow2-uncompressed]
+name=Phony Fedora qcow2 uncompressed
+file=fedora.qcow2
+format=qcow2
+size=1073741824
+expand=/dev/sda2
+lvexpand=/dev/VG/Root
+notes=Phony Fedora look-alike used for testing.
+
+[phony-fedora-no-format]
+name=Phony Fedora
+file=fedora.qcow2.xz
+size=1073741824
+expand=/dev/sda2
+lvexpand=/dev/VG/Root
+notes=Phony Fedora look-alike used for testing.
+
 [phony-ubuntu]
 name=Phony Ubuntu
 file=ubuntu.xz
diff --git a/builder/test-virt-builder-planner.sh b/builder/test-virt-builder-planner.sh
new file mode 100755
index 0000000..738b299
--- /dev/null
+++ b/builder/test-virt-builder-planner.sh
@@ -0,0 +1,51 @@
+#!/bin/bash -
+# libguestfs virt-builder test script
+# Copyright (C) 2013 Red Hat Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+export LANG=C
+set -e
+
+abs_builddir=$(pwd)
+
+export VIRT_BUILDER_SOURCE=file://$abs_builddir/test-index
+
+if [ ! -f fedora.xz -o ! -f fedora.qcow2 -o ! -f fedora.qcow2.xz ]; then
+    echo "$0: test skipped because there is no fedora.xz, fedora.qcow2 or fedora.qcow2.xz in the build directory"
+    exit 77
+fi
+
+if [ "$(../fish/guestfish get-backend)" = "uml" ]; then
+    echo "$0: test skipped because backend is UML"
+    exit 77
+fi
+
+rm -f planner-output
+
+for input in phony-fedora phony-fedora-qcow2 phony-fedora-qcow2-uncompressed phony-fedora-no-format; do
+    for size in none 1G 1.1G 2G; do
+        for format in none raw qcow2; do
+            args="--output planner-output --no-cache --no-check-signature"
+            if [ "$size" != "none" ]; then args="$args --size $size"; fi
+            if [ "$format" != "none" ]; then args="$args --format $format"; fi
+
+            echo $VG ./virt-builder $input $args
+            $VG ./virt-builder $input $args
+        done
+    done
+done
+
+rm planner-output
-- 
1.8.4.2

